File: ceed/editors/looknfeel/hierarchy_tree_item.py
##############################################################################
#   created:    25th June 2014
#   author:     Lukas E Meindl
##############################################################################
##############################################################################
#   CEED - Unified CEGUI asset editor
#
#   Copyright (C) 2011-2014   Martin Preisler <martin@preisler.me>
#                             and contributing authors (see AUTHORS file)
#
#   This program is free software: you can redistribute it and/or modify
#   it under the terms of the GNU General Public License as published by
#   the Free Software Foundation, either version 3 of the License, or
#   (at your option) any later version.
#
#   This program is distributed in the hope that it will be useful,
#   but WITHOUT ANY WARRANTY; without even the implied warranty of
#   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#   GNU General Public License for more details.
#
#   You should have received a copy of the GNU General Public License
#   along with this program.  If not, see <http://www.gnu.org/licenses/>.
##############################################################################

from PySide import QtCore
from PySide import QtGui

import PyCEGUI


class LookNFeelHierarchyItem(QtGui.QStandardItem):

    def __init__(self, falagardElement, prefix=""):
        """
        Creates a hierarchy item based on an element of the WidgetLookFeel object. The element can be
        WidgetLookFeel itself or any child node (such as StateImagery or NamedArea, etc.) that is contained
        in the WidgetLookFeel or any of its children. The children of this element will created as well. The
        resulting hierarchy will be equal, or at least very similar, to the node hierarchy seen in the XML file.
        :param falagardElement:
        :return:
        """

        self.falagardElement = falagardElement
        self.prefix = prefix

        name, toolTip = self.getNameAndToolTip(falagardElement, self.prefix)

        super(LookNFeelHierarchyItem, self).__init__(name)
        self.setToolTip(toolTip)

        self.setFlags(QtCore.Qt.ItemIsEnabled |
                      QtCore.Qt.ItemIsSelectable)

        self.setData(None, QtCore.Qt.CheckStateRole)

        self.createChildren()

    @staticmethod
    def getNameAndToolTip(falagardElement, prefix):
        """
        Creates a name and tooltip for any element that can be part of a WidgetLookFeel and returns it
        :param falagardElement:
        :return: str, str
        """
        name = prefix

        from ceed.editors.looknfeel.tabbed_editor import LookNFeelTabbedEditor
        toolTip = u"type: " + LookNFeelTabbedEditor.getFalagardElementTypeAsString(falagardElement)

        # The WidgetLookFeel element:
        if isinstance(falagardElement, PyCEGUI.WidgetLookFeel):
            name += "Properties, PropertyDefinitions, PropertyLinkDefinitions"

        # Elements that can be children of a WidgetLookFeel:
        elif isinstance(falagardElement, PyCEGUI.NamedArea):
            name += u"NamedArea: \"" + falagardElement.getName() + u"\""
        elif isinstance(falagardElement, PyCEGUI.ImagerySection):
            name += u"ImagerySection: \"" + falagardElement.getName() + u"\""
        elif isinstance(falagardElement, PyCEGUI.StateImagery):
            name += u"StateImagery: \"" + falagardElement.getName() + u"\""
            toolTip = u"type: StateImagery"
        elif isinstance(falagardElement, PyCEGUI.WidgetComponent):
            name += u"Child: \"" + falagardElement.getWidgetName() + u"\" type: " + falagardElement.getWidgetLookName() + u""

        # Elements that can be children of a ImagerySection:
        elif isinstance(falagardElement, PyCEGUI.ImageryComponent):
            name += u"ImageryComponent"
        elif isinstance(falagardElement, PyCEGUI.TextComponent):
            name += u"TextComponent"
        elif isinstance(falagardElement, PyCEGUI.FrameComponent):
            name += u"FrameComponent"

        # Elements that can be children of a StateImagery:
        elif isinstance(falagardElement, PyCEGUI.LayerSpecification):
            name += u"Layer"

        # Elements that can be children of a LayerSpecification:
        elif isinstance(falagardElement, PyCEGUI.SectionSpecification):
            name += u"Section: \"" + falagardElement.getSectionName() + "\""

        # The ComponentArea element
        elif isinstance(falagardElement, PyCEGUI.ComponentArea):
            name += u"Area"

        # The Image element
        elif isinstance(falagardElement, PyCEGUI.Image):
            name += u" name:\"" + falagardElement.getName() + u"\""

        return name, toolTip

    def createChildren(self):
        """
        Creates the children items for for the this item's WidgetLookFeel element
        :param:
        :return:
        """
        if isinstance(self.falagardElement, PyCEGUI.NamedArea):
            self.createNamedAreaChildren()
        elif isinstance(self.falagardElement, PyCEGUI.ImagerySection):
            self.createImagerySectionChildren()
        elif isinstance(self.falagardElement, PyCEGUI.StateImagery):
            self.createStateImageryChildren()
        elif isinstance(self.falagardElement, PyCEGUI.WidgetComponent):
            self.createWidgetComponentChildren()
        elif isinstance(self.falagardElement, PyCEGUI.ImageryComponent):
            self.createImageryComponentChildren()
        elif isinstance(self.falagardElement, PyCEGUI.TextComponent):
            self.createTextComponentChildren()
        elif isinstance(self.falagardElement, PyCEGUI.FrameComponent):
            self.createFrameComponentChildren()
        elif isinstance(self.falagardElement, PyCEGUI.LayerSpecification):
            self.createLayerSpecificationChildren()
        elif isinstance(self.falagardElement, PyCEGUI.SectionSpecification):
            self.createSectionSpecificationChildren()

    def createAndAddItem(self, falagardElement, prefix=u""):
        """
        Creates an item based on the supplied object and adds it to this object. A prefix string can be added to the displayed name.
        :param falagardElement:
        :param prefix: str
        :return:
        """
        newItem = LookNFeelHierarchyItem(falagardElement, prefix)
        self.appendRow(newItem)

    def createNamedAreaChildren(self):
        """
        Creates and appends an item for the Area of the NamedArea
        :return:
        """
        area = self.falagardElement.getArea()
        self.createAndAddItem(area)

    def createImagerySectionChildren(self):
        """
        Creates and appends children items based on an ImagerySection.
        :return:
        """
        frameCompIter = self.falagardElement.getFrameComponentIterator()
        while not frameCompIter.isAtEnd():
            currentFrameComp = frameCompIter.getCurrentValue()
            self.createAndAddItem(currentFrameComp)
            frameCompIter.next()

        textCompIter = self.falagardElement.getTextComponentIterator()
        while not textCompIter.isAtEnd():
            currentTextComp = textCompIter.getCurrentValue()
            self.createAndAddItem(currentTextComp)
            textCompIter.next()

        imageryCompIter = self.falagardElement.getImageryComponentIterator()
        while not imageryCompIter.isAtEnd():
            imageryComp = imageryCompIter.getCurrentValue()
            self.createAndAddItem(imageryComp)
            imageryCompIter.next()

    def createStateImageryChildren(self):
        """
        Creates and appends children items based on an ImagerySection.
        :return:
        """
        layerIter = self.falagardElement.getLayerIterator()
        while not layerIter.isAtEnd():
            layer = layerIter.getCurrentValue()
            self.createAndAddItem(layer)
            layerIter.next()

    def createWidgetComponentChildren(self):
        """
        Creates and appends children items based on a WidgetComponent (child widget).
        :return:
        """
        area = self.falagardElement.getComponentArea()
        self.createAndAddItem(area)

    def createImageryComponentChildren(self):
        """
        Creates and appends children items based on an ImageryComponent.
        :return:
        """
        area = self.falagardElement.getComponentArea()
        self.createAndAddItem(area)

    def createTextComponentChildren(self):
        """
        Creates and appends children items based on a TextComponent.
        :return:
        """
        area = self.falagardElement.getComponentArea()
        self.createAndAddItem(area)

    def createFrameComponentChildren(self):
        """
        Creates and appends children items based on a FrameComponent.
        :return:
        """
        area = self.falagardElement.getComponentArea()
        self.createAndAddItem(area)

        testWindow = PyCEGUI.WindowManager.getSingleton().createWindow("DefaultWindow", "")

        image = self.falagardElement.getImage(PyCEGUI.FrameImageComponent.FIC_TOP_LEFT_CORNER, testWindow)
        if image is not None:
            self.createAndAddItem(image, u"Image: TopLeftCorner")
        image = self.falagardElement.getImage(PyCEGUI.FrameImageComponent.FIC_TOP_RIGHT_CORNER, testWindow)
        if image is not None:
            self.createAndAddItem(image, u"Image: TopRightCorner")
        image = self.falagardElement.getImage(PyCEGUI.FrameImageComponent.FIC_BOTTOM_LEFT_CORNER, testWindow)
        if image is not None:
            self.createAndAddItem(image, u"Image: BottomLeftCorner")
        image = self.falagardElement.getImage(PyCEGUI.FrameImageComponent.FIC_BOTTOM_RIGHT_CORNER, testWindow)
        if image is not None:
            self.createAndAddItem(image, u"Image: BottomRightCorner")
        image = self.falagardElement.getImage(PyCEGUI.FrameImageComponent.FIC_LEFT_EDGE, testWindow)
        if image is not None:
            self.createAndAddItem(image, u"Image: LeftEdge")
        image = self.falagardElement.getImage(PyCEGUI.FrameImageComponent.FIC_RIGHT_EDGE, testWindow)
        if image is not None:
            self.createAndAddItem(image, u"Image: RightEdge")
        image = self.falagardElement.getImage(PyCEGUI.FrameImageComponent.FIC_TOP_EDGE, testWindow)
        if image is not None:
            self.createAndAddItem(image, u"Image: TopEdge")
        image = self.falagardElement.getImage(PyCEGUI.FrameImageComponent.FIC_BOTTOM_EDGE, testWindow)
        if image is not None:
            self.createAndAddItem(image, u"Image: BottomEdge")
        image = self.falagardElement.getImage(PyCEGUI.FrameImageComponent.FIC_BACKGROUND, testWindow)
        if image is not None:
            self.createAndAddItem(image, u"Image: Background")

    def createLayerSpecificationChildren(self):
        """
        Creates and appends children items based on a LayerSpecification.
        :return:
        """
        sectionIter = self.falagardElement.getSectionIterator()
        while not sectionIter.isAtEnd():
            currentSectionSpecification = sectionIter.getCurrentValue()
            self.createAndAddItem(currentSectionSpecification)
            sectionIter.next()

    def createSectionSpecificationChildren(self):
        """
        Creates and appends children items based on a SectionSpecification.
        :return:
        """

    def clone(self):
        ret = LookNFeelHierarchyItem(self.manipulator)
        ret.setData(self.data(QtCore.Qt.CheckStateRole), QtCore.Qt.CheckStateRole)
        return ret

File: ceed/src/editors/looknfeel/hierarchy_tree_item.cpp
/*
   CEED - Unified CEGUI asset editor

   Copyright (C) 2011-2017   Martin Preisler <martin@preisler.me>
                             and contributing authors (see AUTHORS file)

   This program is free software: you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation, either version 3 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#include "editors/looknfeel/hierarchy_tree_item.h"

#include "editors/looknfeel/falagard_element_interface.h"

namespace CEED {
namespace editors {
namespace looknfeel {
namespace hierarchy_tree_item {

LookNFeelHierarchyItem::LookNFeelHierarchyItem(FalagardElement falagardElement, const QString &prefix)
    : QStandardItem()
{
    m_falagardElement = falagardElement;
    m_prefix = prefix;

    auto nameAndToolTip = getNameAndToolTip(falagardElement, m_prefix);

    setText(nameAndToolTip.first);
    setToolTip(nameAndToolTip.second);

    setFlags(Qt::ItemIsEnabled |
             Qt::ItemIsSelectable);

    setData(QVariant(), Qt::CheckStateRole);

    createChildren();
}

#if 0
QPair<QString, QString> LookNFeelHierarchyItem::getNameAndToolTip(FalagardElement falagardElement, const QString& prefix)
{
    switch (falagardElement.type()) {
    case FalagardElement::Type::PropertyDefinitionBase:
        return getNameAndToolTip(falagardElement.toPropertyDefinitionBase(), prefix);
    case FalagardElement::Type::PropertyInitialiser:
        return getNameAndToolTip(falagardElement.toPropertyInitialiser(), prefix);
    case FalagardElement::Type::NamedArea:
        return getNameAndToolTip(falagardElement.toNamedArea(), prefix);
    case FalagardElement::Type::ImagerySection:
        return getNameAndToolTip(falagardElement.toImagerySection(), prefix);
    case FalagardElement::Type::StateImagery:
        return getNameAndToolTip(falagardElement.toStateImagery(), prefix);
    case FalagardElement::Type::WidgetComponent:
        return getNameAndToolTip(falagardElement.toWidgetComponent(), prefix);
    case FalagardElement::Type::ImageryComponent:
        return getNameAndToolTip(falagardElement.toImageryComponent(), prefix);
    case FalagardElement::Type::TextComponent:
        return getNameAndToolTip(falagardElement.toTextComponent(), prefix);
    case FalagardElement::Type::FrameComponent:
        return getNameAndToolTip(falagardElement.toFrameComponent(), prefix);
    case FalagardElement::Type::LayerSpecification:
        return getNameAndToolTip(falagardElement.toLayerSpecification(), prefix);
    case FalagardElement::Type::SectionSpecification:
        return getNameAndToolTip(falagardElement.toSectionSpecification(), prefix);
    case FalagardElement::Type::ComponentArea:
        return getNameAndToolTip(falagardElement.toComponentArea(), prefix);
    case FalagardElement::Type::Image:
        return getNameAndToolTip(falagardElement.toImage(), prefix);
    default:
        Q_ASSERT(false);
    }
}

QPair<QString, QString> LookNFeelHierarchyItem::getNameAndToolTip(CEGUI::WidgetLookFeel *falagardElement_, const QString &prefix)
{
    CEGUI::String name = prefix;
    QString toolTip = "type: " + falagard_element_interface::FalagardElementInterface::getFalagardElementTypeAsString(falagardElement_);
    return name, toolTip;
}

QPair<QString, QString> LookNFeelHierarchyItem::getNameAndToolTip(CEGUI::PropertyDefinitionBase *falagardElement, const QString &prefix)
{
    CEGUI::String name = prefix;
    name += "\"" + falagardElement->getPropertyName() + "\"";
    QString toolTip = "type: " + falagard_element_interface::FalagardElementInterface::getFalagardElementTypeAsString(falagardElement_);
    return name, toolTip;
}

QPair<QString, QString> LookNFeelHierarchyItem::getNameAndToolTip(CEGUI::PropertyInitialiser *falagardElement, const QString &prefix)
{
    CEGUI::String name = prefix;
    name += "\"" + falagardElement->getTargetPropertyName() + "\"";
    QString toolTip = "type: " + falagard_element_interface::FalagardElementInterface::getFalagardElementTypeAsString(falagardElement_);
    return name, toolTip;
}

QPair<QString, QString> LookNFeelHierarchyItem::getNameAndToolTip(CEGUI::NamedArea *falagardElement, const QString &prefix)
{
    CEGUI::String name = prefix;
    name += "\"" + falagardElement->getName() + "\"";
    QString toolTip = "type: " + falagard_element_interface::FalagardElementInterface::getFalagardElementTypeAsString(falagardElement_);
    return name, toolTip;
}

QPair<QString, QString> LookNFeelHierarchyItem::getNameAndToolTip(CEGUI::ImagerySection *falagardElement, const QString &prefix)
{
    CEGUI::String name = prefix;
    name += "\"" + falagardElement->getName() + "\"";
    QString toolTip = "type: " + falagard_element_interface::FalagardElementInterface::getFalagardElementTypeAsString(falagardElement_);
    return name, toolTip;
}

QPair<QString, QString> LookNFeelHierarchyItem::getNameAndToolTip(CEGUI::StateImagery *falagardElement, const QString &prefix)
{
    CEGUI::String name = prefix;
    name += "\"" + falagardElement->getName() + "\"";
    QString toolTip = "type: " + falagard_element_interface::FalagardElementInterface::getFalagardElementTypeAsString(falagardElement_);
    return name, toolTip;
}

QPair<QString, QString> LookNFeelHierarchyItem::getNameAndToolTip(CEGUI::WidgetComponent *falagardElement, const QString &prefix)
{
    CEGUI::String name = prefix;
    name += "\"" + falagardElement->getWidgetName() + "\"" + " (\"" + falagardElement->getBaseWidgetType() + "\")";
    QString toolTip = "type: " + falagard_element_interface::FalagardElementInterface::getFalagardElementTypeAsString(falagardElement_);
    return name, toolTip;
}

QPair<QString, QString> LookNFeelHierarchyItem::getNameAndToolTip(CEGUI::ImageryComponent *falagardElement, const QString &prefix)
{
    CEGUI::String name = prefix;
    name += "ImageryComponent";
    QString toolTip = "type: " + falagard_element_interface::FalagardElementInterface::getFalagardElementTypeAsString(falagardElement_);
    return name, toolTip;
}

QPair<QString, QString> LookNFeelHierarchyItem::getNameAndToolTip(CEGUI::TextComponent *falagardElement, const QString &prefix)
{
    CEGUI::String name = prefix;
    name += "TextComponent";
    QString toolTip = "type: " + falagard_element_interface::FalagardElementInterface::getFalagardElementTypeAsString(falagardElement_);
    return name, toolTip;
}

QPair<QString, QString> LookNFeelHierarchyItem::getNameAndToolTip(CEGUI::FrameComponent *falagardElement, const QString &prefix)
{
    CEGUI::String name = prefix;
    name += "FrameComponent";
    QString toolTip = "type: " + falagard_element_interface::FalagardElementInterface::getFalagardElementTypeAsString(falagardElement_);
    return name, toolTip;
}

QPair<QString, QString> LookNFeelHierarchyItem::getNameAndToolTip(CEGUI::LayerSpecification *falagardElement, const QString &prefix)
{
    CEGUI::String name = prefix;
    name += "Layer";
    QString toolTip = "type: " + falagard_element_interface::FalagardElementInterface::getFalagardElementTypeAsString(falagardElement_);
    return name, toolTip;
}

QPair<QString, QString> LookNFeelHierarchyItem::getNameAndToolTip(CEGUI::SectionSpecification *falagardElement, const QString &prefix)
{
    CEGUI::String name = prefix;
    name += "Section: \"" + falagardElement->getSectionName() + "\"";
    QString toolTip = "type: " + falagard_element_interface::FalagardElementInterface::getFalagardElementTypeAsString(falagardElement_);
    return name, toolTip;
}

QPair<QString, QString> LookNFeelHierarchyItem::getNameAndToolTip(CEGUI::ComponentArea *falagardElement, const QString &prefix)
{
    CEGUI::String name = prefix;
    name += "Area";
    QString toolTip = "type: " + falagard_element_interface::FalagardElementInterface::getFalagardElementTypeAsString(falagardElement_);
    return name, toolTip;
}

QPair<QString, QString> LookNFeelHierarchyItem::getNameAndToolTip(CEGUI::Image *falagardElement, const QString &prefix)
{
    CEGUI::String name = prefix;
    name += " \"" + falagardElement->getName() + "\"";
    QString toolTip = "type: " + falagard_element_interface::FalagardElementInterface::getFalagardElementTypeAsString(falagardElement_);
    return name, toolTip;
}

#else

QPair<QString, QString> LookNFeelHierarchyItem::getNameAndToolTip(FalagardElement falagardElement_, const QString& prefix)
{
    QString name = prefix;

    QString toolTip = "type: " + falagard_element_interface::FalagardElementInterface::getFalagardElementTypeAsString(falagardElement_);

    if (auto falagardElement = falagardElement_.toPropertyDefinitionBase()) {
        name += "\"" + TO_QSTR(falagardElement->getPropertyName()) + "\"";

    } else if (auto falagardElement = falagardElement_.toPropertyInitialiser()) {
        name += "\"" + TO_QSTR(falagardElement->getTargetPropertyName()) + "\"";

    } else if (auto falagardElement = falagardElement_.toNamedArea()) {
        name += "\"" + TO_QSTR(falagardElement->getName()) + "\"";
    } else if (auto falagardElement = falagardElement_.toImagerySection()) {
        name += "\"" + TO_QSTR(falagardElement->getName()) + "\"";
    } else if (auto falagardElement = falagardElement_.toStateImagery()) {
        name += "\"" + TO_QSTR(falagardElement->getName()) + "\"";
    } else if (auto falagardElement = falagardElement_.toWidgetComponent()) {
        name += "\"" + TO_QSTR(falagardElement->getWidgetName()) + "\"" + " (\"" + TO_QSTR(falagardElement->getBaseWidgetType()) + "\")";

    // Elements that can be children of a ImagerySection:
    } else if (auto falagardElement = falagardElement_.toImageryComponent()) {
        name += "ImageryComponent";
    } else if (auto falagardElement = falagardElement_.toTextComponent()) {
        name += "TextComponent";
    } else if (auto falagardElement = falagardElement_.toFrameComponent()) {
        name += "FrameComponent";

    // Elements that can be children of a StateImagery:
    } else if (auto falagardElement = falagardElement_.toLayerSpecification()) {
        name += "Layer";

    // Elements that can be children of a LayerSpecification:
    } else if (auto falagardElement = falagardElement_.toSectionSpecification()) {
        name += "Section: \"" + TO_QSTR(falagardElement->getSectionName()) + "\"";

    // The ComponentArea element
    } else if (auto falagardElement = falagardElement_.toComponentArea()) {
        name += "Area";

    // The Image element
    } else if (auto falagardElement = falagardElement_.toImage()) {
        name += " \"" + TO_QSTR(falagardElement->getName()) + "\"";
    }

    return { name, toolTip };
}
#endif

void LookNFeelHierarchyItem::createChildren()
{
    if (auto falagardElement = m_falagardElement.toNamedArea()) {
        createNamedAreaChildren(falagardElement);
    } else if (auto falagardElement = m_falagardElement.toImagerySection()) {
        createImagerySectionChildren(falagardElement);
    } else if (auto falagardElement = m_falagardElement.toStateImagery()) {
        createStateImageryChildren(falagardElement);
    } else if (auto falagardElement = m_falagardElement.toWidgetComponent()) {
        createWidgetComponentChildren(falagardElement);
    } else if (auto falagardElement = m_falagardElement.toImageryComponent()) {
        createImageryComponentChildren(falagardElement);
    } else if (auto falagardElement = m_falagardElement.toTextComponent()) {
        createTextComponentChildren(falagardElement);
    } else if (auto falagardElement = m_falagardElement.toFrameComponent()) {
        createFrameComponentChildren(falagardElement);
    } else if (auto falagardElement = m_falagardElement.toLayerSpecification()) {
        createLayerSpecificationChildren(falagardElement);
    }
}

void LookNFeelHierarchyItem::createAndAddItem(FalagardElement falagardElement, const QString &prefix)
{
    auto newItem = new LookNFeelHierarchyItem(falagardElement, prefix);
    appendRow(newItem);
}

void LookNFeelHierarchyItem::createNamedAreaChildren(CEGUI::NamedArea* falagardElement)
{
    auto& area = falagardElement->getArea();
    createAndAddItem(&area);
}

void LookNFeelHierarchyItem::createImagerySectionChildren(CEGUI::ImagerySection* falagardElement)
{
    auto frameComponentList = falagardElement->getFrameComponentPointers();
    for (auto frameComponent : frameComponentList)
        createAndAddItem(frameComponent);

    auto textComponentList = falagardElement->getTextComponentPointers();
    for (auto textComponent : textComponentList)
        createAndAddItem(textComponent);

    auto imageryComponentList = falagardElement->getImageryComponentPointers();
    for (auto imageryComponent : imageryComponentList)
        createAndAddItem(imageryComponent);
}

void LookNFeelHierarchyItem::createStateImageryChildren(CEGUI::StateImagery* falagardElement)
{
    auto layerSpecList = falagardElement->getLayerSpecificationPointers();
    for (auto layerSpec : layerSpecList)
        createAndAddItem(layerSpec);
}

void LookNFeelHierarchyItem::createWidgetComponentChildren(CEGUI::WidgetComponent* falagardElement)
{
    auto& area = falagardElement->getComponentArea();
    createAndAddItem(&area);
}

void LookNFeelHierarchyItem::createImageryComponentChildren(CEGUI::ImageryComponent* falagardElement)
{
    auto& area = falagardElement->getComponentArea();
    createAndAddItem(&area);
}

void LookNFeelHierarchyItem::createTextComponentChildren(CEGUI::TextComponent* falagardElement)
{
    auto& area = falagardElement->getComponentArea();
    createAndAddItem(&area);
}

void LookNFeelHierarchyItem::createFrameComponentChildren(CEGUI::FrameComponent *falagardElement)
{
    auto& area = falagardElement->getComponentArea();
    createAndAddItem(&area);
}

void LookNFeelHierarchyItem::createLayerSpecificationChildren(CEGUI::LayerSpecification *falagardElement)
{
    auto sectionSpecList = falagardElement->getSectionSpecificationPointers();
    for (auto sectionSpec : sectionSpecList)
        createAndAddItem(sectionSpec);
}


} // namespace hierarchy_tree_item
} // namespace looknfeel
} // namespace editors
} // namespace CEED
